dali/test/op_graph.h

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_OP_GRAPH_H_
+#define DALI_TEST_OP_GRAPH_H_
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <iostream>
+
+namespace dali {
+namespace testing {
+
+struct Node;
+
+struct Edge {
+  Edge() = default;
+  Edge(Node *n, int index, bool explicit_GPU = false) : node(n), index(index), explicit_GPU(false) {}
+
+  Node *node = nullptr;
+  int index = 0;
+  bool explicit_GPU = false;
+
+  Edge gpu() const {
+    auto ret = *this;
+    ret.explicit_GPU = true;
+    return ret;
+  }
+};
+
+struct Node {
+  Node() = default;
+  Node(const Node &) = delete;
+  Node(Node &&) = default;
+
+  Node(std::string op_name, std::string node_name, int backend_mask)
+  : op_name(std::move(op_name)), node_name(std::move(node_name)), backend_mask(backend_mask) {}
+
+  template <typename... Inputs>
+  Node &operator()(Inputs&&... inputs) {
+    int junk[] = { (add_in(inputs), 0)... };
+    (void)junk;
+    return *this;
+  }
+
+  void add_in(Edge e) {
+    in.push_back(e);
+    if (e.index >= (int)e.node->out.size())
+      e.node->out.resize(e.index+1);
+    e.node->out[e.index] = Edge(this, in.size()-1);
+  }
+
+  Edge operator[](int i) {
+    if (i >= num_outputs) num_outputs = i+1;
+    return { this, i };
+  }
+
+  enum Backend {
+    CPU = 1,
+    GPU = 2,
+    AnyBackend = 3,
+  };
+
+  std::string op_name;
+  std::string node_name;
+  int backend_mask = 0;
+  int num_outputs = 0;
+  std::vector<Edge> in;
+  std::vector<Edge> out;
+};
+
+struct OpDAG {
+  Node &add(std::string op_name, std::string node_name, int backend_mask = Node::AnyBackend) {
```

Comment:
According to Google Style (which we use), [this should be a](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) `class`. Also, [the name should be](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) `OpDag` (I'd prefer `OperatorDag` though)

Reply:
> < rant >I really do hate the practice of decapitalizing acronyms. < /rant >

I don't wanna be this kind of guy that pick holes or anything ;) It's just in my opinion, there's a purpose, why we agreed to use the Google Style - consistency. I'm not saying, that it's unchangeable, it's just the whole point of consistency is that everybody writes a code in the same way. Anywho, I'm totally open for any discussion for code guidelines :)

Reply:
Making it a `class` would call for the ridiculous underscore-suffixed members - and possibly making them private.
I want to have public `in` and `out` in the DAG object. However, I can use a NOLINT for that...
Making `in` and `out` functions doesn't play well with the rest of the syntax when we use both `operator[]` for extracting outputs  and `operator()` for assigning inputs. Actually, `out` could be a function, but `in` cannot - and I'd rather keep them consistent.